### PR TITLE
Fix typo in apimachinery

### PR DESCRIPTION
**Pull request by hezhizhen**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Fix a typo in apimachinery/pkg/util/runtime/runtime.go

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by hezhizhen** [comment]
I signed it

**Comment by hezhizhen** [comment]
@ahmad-diaa what else should I do?

**Comment by liggitt** [comment]
This fixes a lint error so now you get to remove this package from the lint failures list:
```
Some packages in hack/.golint_failures are passing golint. Please remove them.
  staging/src/k8s.io/apimachinery/pkg/util/runtime
```

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ var ErrorHandlers = []func(error){
 	}).OnError,
 }
 
-// HandlerError is a method to invoke when a non-user facing piece of code cannot
+// HandleError is a method to invoke when a non-user facing piece of code cannot
 // return an error and needs to indicate it has been ignored. Invoking this method
```
Comment (ahmad-diaa) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: I think `to indicate it has been ignored` might be confusing. `to indicate that the/this error has been ignored` could be clearer. 
We can change it as well. @liggitt WDYT?

**Comment by hezhizhen** [comment]
> This fixes a lint error so now you get to remove this package from the lint failures list:
> 
> ```
> Some packages in hack/.golint_failures are passing golint. Please remove them.
>   staging/src/k8s.io/apimachinery/pkg/util/runtime
> ```

Thank you

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go (hunk 1 above)
Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: Let me re-phrase it.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 2]
```diff
@@ -94,12 +94,17 @@ var ErrorHandlers = []func(error){
...
 // is preferable to logging the error - the default behavior is to log but the
 // errors may be sent to a remote server for analysis.
+
+// HandleError handles an error by a list of pre-defined error handlers, including
+// logging and sending to a remote server for analysis, and thus calling this
+// function is preferablt to logging the error. It is for the case when the error
```
Comment (hezhizhen) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: preferablt -> preferable

**Comment by nikhita** [comment]
@hezhizhen thanks for your PR! Can you also squash the commits to a single commit?

Also, the test failures look like flakes and are not related to your PR.

**Comment by hezhizhen** [comment]
> @hezhizhen thanks for your PR! Can you also squash the commits to a single commit?
> 
> Also, the test failures look like flakes and are not related to your PR.

OK, I'll squash all the commits when the PR looks good to all.

**Comment by nikhita** [comment]
The job is broken right now so retesting it won't help: https://github.com/kubernetes/kubernetes/issues/83086 :(

Adding a hold to avoid retesting, until the issue has been fixed.
/hold

**Comment by nikhita** [comment]
Oh oops, I thought this PR already had lgtm and approved labels. :grimacing: 

/hold cancel

**Comment by hezhizhen** [comment]
any comments? @ahmad-diaa @liggitt

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 3]
```diff
@@ -94,12 +94,12 @@ var ErrorHandlers = []func(error){
...
 }
 
-// HandlerError is a method to invoke when a non-user facing piece of code cannot
-// return an error and needs to indicate it has been ignored. Invoking this method
-// is preferable to logging the error - the default behavior is to log but the
-// errors may be sent to a remote server for analysis.
+// HandleError handles an error by a list of pre-defined error handlers, including
+// logging and sending to a remote server for analysis, and thus calling this
+// function is preferable to logging the error. It is for the case when the error
+// can't be returned or should be ignored.
```
Comment (ahmad-diaa) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: I think `and has been ignored` matches the previous comment more than `should be ignored`

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go [hunk 4]
```diff
@@ -144,15 +144,20 @@ func captureStderr(f func()) (string, error) {
 	defer func() { os.Stderr = bak }()
 
 	resultCh := make(chan string)
+	errCh := make(chan error)
```
Comment (ahmad-diaa) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: I am not really familiar with the code here, but why is this getting changed?

Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: Because the `io.Copy` function returns an error and it was not checked before.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go (hunk 3 above)
Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: done. PTAL

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: sending to two channels in order and receiving from two channels in order in a return is a bit hard to follow.

suggest this instead:

```go
var copyErr error
// copy the output in a separate goroutine so printing can't block indefinitely
go func() {
  ...
  _, copyErr = io.Copy(&buf, r)
  ...
}
...
return <-resultCh, copyErr
```

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go [hunk 5]
```diff
@@ -144,15 +144,20 @@ func captureStderr(f func()) (string, error) {
...
 	// copy the output in a separate goroutine so printing can't block indefinitely
 	go func() {
 		var buf bytes.Buffer
-		io.Copy(&buf, r)
+		_, copyErr := io.Copy(&buf, r)
```
Comment (ahmad-diaa) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: s/copyErr/err, more consistent with the code.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: I'd stick with `copyErr` to make it clear where the error came from. Also, this is masking the variable:

```suggestion
		_, copyErr = io.Copy(&buf, r)
```

Comment (ahmad-diaa) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: Makes sense if the code is changed to use var instead of `errCh` as you pointed out earlier. If not, the error is being sent to the channel in the following line and `err` would be sufficient.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go (hunk 4 above)
Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: Why should sending and receiving be in order? I think the order doesn't matter and the function returns when both of the channels have received values.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go (hunk 5 above)
Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: Agree with @liggitt . showing error's source is better no matter if it is copied to an outside variable or sent to a channel.

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: > Why should sending and receiving be in order?

return values are evaluated in order, so you are depending on the channel send order inside the goroutine to be the same as the channel receive order in the return statement. That happens to work in your example, but is a slightly weird construction.

https://play.golang.org/p/dbj85uV4bDX is an example of the deadlock that occurs if the send and receive order did not match

Setting the error outside the closure and simply returning that seems more straightforward

Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: Thanks for your explanation, I didn't even know it.

Comment (hezhizhen) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime_test.go]: done. PTAL @liggitt @ahmad-diaa

**Comment by hezhizhen** [comment]
any comments? @ahmad-diaa @liggitt

File: staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go [hunk 6]
```diff
@@ -94,12 +94,12 @@ var ErrorHandlers = []func(error){
 	}).OnError,
 }
 
-// HandlerError is a method to invoke when a non-user facing piece of code cannot
-// return an error and needs to indicate it has been ignored. Invoking this method
-// is preferable to logging the error - the default behavior is to log but the
-// errors may be sent to a remote server for analysis.
+// HandleError handles an error by a list of pre-defined error handlers, including
+// logging and sending to a remote server for analysis, and thus calling this
```
Comment (ahmad-diaa) [inline staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: Sending the error to a remote server is not included in `ErrorHandlers`, that's why the old comment was more accurate regarding this part `may be sent to a remote server for analysis`

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go]: agree
